vermouth/processors/apply_rubber_band.py

```diff
@@ -158,30 +163,24 @@ def build_connectivity_matrix(graph, separation, selected_nodes):
     numpy.ndarray
         A boolean matrix.
     """
-    if separation < 0:
-        raise ValueError('Separation has to be zero or positive.')
-    if separation == 0:
-        # The connectivity matrix with a separation of 0 is the adjacency
-        # matrix. Thankfully, networkx can directly give it to us a a numpy
-        # matrix.
-        return np.asarray(nx.to_numpy_matrix(graph, nodelist=selected_nodes).astype(bool))
+    res_graph = make_residue_graph(graph)
+    distance_pairs = nx.all_pairs_shortest_path_length(res_graph, cutoff=separation)
+    # only gets "positive" entries due to the cutoff argument above
     size = graph.number_of_nodes()
-    correspondence = {node: index for index, node in enumerate(graph.nodes)}
     # the matrix will be reduced before returning it but nx.all_pairs_shortest_path_length
     # does not take a subset of nodes
     connectivity = np.zeros((size, size), dtype=bool)
-    # separation is provided in term of nodes while the path is provided in
-    # terms of edges, hence `separation + 1`.
-    distance_pairs = nx.all_pairs_shortest_path_length(graph, cutoff=separation + 1)
-    #only gets "positive" entries due to the cutoff argument above
-    for origin, target_and_distances in distance_pairs:
-        for target in target_and_distances:
-            connectivity[correspondence[origin], correspondence[target]] = True
+    for origin_residue, matchs_distances in distance_pairs:
+        for target_residue in matchs_distances:
+            origin_nodes = res_graph.nodes[origin_residue]['graph'].nodes()
+            target_nodes = res_graph.nodes[target_residue]['graph'].nodes()
+            for origin, target in itertools.product(origin_nodes, target_nodes):
+                connectivity[node_to_idx[origin], node_to_idx[target]] = True
     np.fill_diagonal(connectivity, False)
```

Comment:
You will still get elastic bonds within a residue (but not from a bead to itself). `all_pairs_shortest_path_length` will return the item `{1: {1: 0}}` as well.
Is this the desired behaviour?

Reply:
@pckroon  yes it does return those but these are the ones on the diagonal and that is set to False after in line 179

Reply:
No, they're not :)

The numbers returned by `all_pairs_shortest_path_length` are node keys in the residue graph. The diagonal set to False is on a bead level.

Reply:
ahh yes I am stupid didn't think about it. 1) you don't get bonds within a residue, we have tested this a lot; 2) this is the connectivity matrix so things within one residue are connected. It get's inverted and then you have these beads set to false.